src/ui/Distance/DistanceResources.ts

```diff
@@ -0,0 +1,315 @@
+     * 
+     * The default value is `coveo-distance-disabled`.
+     */
+    disabledDistanceCssClass: ComponentOptions.buildStringOption({
+      defaultValue: 'coveo-distance-disabled'
+    }),
+    /**
+     * The latitude to use if no other position was provided.
+     * 
+     * You must also set `longitudeValue` if you specify this value.
+     */
+    latitudeValue: ComponentOptions.buildNumberOption({
+      float: true
+    }),
+    /**
+     * The longitude to use if no other position was provided.
+     * 
+     * You must also set `latitude` if you specify this value.
+     */
+    longitudeValue: ComponentOptions.buildNumberOption({
+      float: true
+    }),
+    /**
+     * The API key to use to request the Google API geolocation service.
+     * 
+     * If not defined, will not try to request the service.
+     */
+    googleApiKey: ComponentOptions.buildStringOption(),
+    /**
+     * Whether to request the browser's geolocation service.
+     * 
+     * If not defined, will not try to request the service.
+     * 
+     * Note that most recent browsers requires your site to be in HTTPS to use its geolocation service.
+     */
+    useNavigator: ComponentOptions.buildBooleanOption(),
+    /**
+     * Whether to execute a new query when a new position has been provided.
+     * 
+     * Default value is `false`.
+     */
+    triggerNewQueryOnNewPosition: ComponentOptions.buildBooleanOption({
+      defaultValue: false
+    })
+  };
+
+  /**
+   * Creates a new `DistanceResources` component.
+   * @param element The HTMLElement on which to instantiate the component.
+   * @param options The options for the `DistanceResources` component.
+   * @param bindings The bindings that the component requires to function normally. If not set, these will be
+   * automatically resolved (with a slower execution time).
+   */
+  constructor(public element: HTMLElement, public options: IDistanceOptions, public bindings: IComponentBindings) {
+    super(element, DistanceResources.ID, bindings);
+
+    this.options = ComponentOptions.initComponentOptions(element, DistanceResources, options);
+    this.bind.onRootElement(InitializationEvents.afterComponentsInitialization, () => this.onAfterComponentsInitialization());
+  }
+
+  /**
+   * Override the current position with the provided values.
+   *
+   * Does not triggers a query automatically.
+   * @param latitude The latitude to set.
+   * @param longitude The longitude to set.
+   */
+  public setPosition(latitude: number, longitude: number): void {
+    this.latitude = latitude;
+    this.longitude = longitude;
+
+    const args: IPositionResolvedEventArgs = {
+      position: {
+        lat: latitude,
+        long: longitude
+      }
+    };
+
+    this.bind.trigger(this.element, DistanceEvents.onPositionResolved, args);
+    this.registerDistanceQuery();
+    if (this.options.triggerNewQueryOnNewPosition) {
```

Comment:
Nice work on the refactor ! 👍 

This is my only remaining concern : 

So to get a valid/working configuration, where you need to know the user location before the first query is done, we would need the following configuration : 

* Disable autoTriggerQuery on the search interface
* triggerNewQueryOnNewPosition=true on the distance component.

This *can* work, but seems kinda strange. It looks very easy to configure badly, so that you only get the @distance facet to show up only on the 2nd query, or the @distance sort to throw errors on the first query etc. 

Ideally, this component would need to "block" the initialization process or any query being triggered before it is able to either resolve a position, or that the user clicks explicitly refuse to give his position.

I think this option should be removed, and I'm sure there's a way to make it work without any option or configuration. As a rule of thumb, I am of the opinion that any option on any component has to be 100% mandatory or extremely straightforward, and if an option exists only to help us simplify our code, that's a failure on our part. 

We should not be pushing complexity to the user configuring the component.

There's a way currently to cancel a query (by setting cancel=true on a `newQuery` event arguments for example), so this component could cancel any query being sent until it is satisfied with his state (either a position was resolved, or the end user refused to give his position), and then launch the query.

What do you guys think @samisayegh  and @dms1lva  ? 
I don't want to be alone in this review :)

Reply:
What about replacing `triggerNewQueryOnNewPosition` with `blockQueryUntilPositionResolved`? 😄

This way, if set to `true`, we cancel all the queries until we get the position.

If set to `false`, the position will be applied only on the next query.

Reply:
I think having `cancelQueriesUntilPositionResolved` as an option makes sense, but as @olamothe said, when the user refuses to give his position, we should execute the query. 

I see two scenarios when using the distance field:

- Search page where showing results without sorting according to the distance field makes no sense.
- Search page where the distance field is used for one of multiple facets.

In the second scenario, if `cancelQueriesUntilPositionResolve` is false, the user might want to do a query as soon as a position is resolved. Especially when using `navigator.geolocation`, when the user clicks allow, it makes sense to make a new query. Should we also keep `triggerNewQueryOnNewPosition` ? Should that be configurable depending on the provider? We might want to wait on the google api but not on the navigator provider.

It might also be good to have a mechanism to have some kind of fallback position provider while we wait for a more accurate one. Even in the first scenario mentioned above, you might want to show _something_ while waiting for the user allowing the geolocation. 


Exposing a method that lets you register a new position provider could also be useful. In that case we would need some kind of mechanism to specify a priority.

I agree with @olamothe on the fact that the most simple configuration should be the one that works and we should choose default values that support that.

Reply:
> In the second scenario, if ... Should that be configurable depending on the provider?
> Exposing a method that lets you register a new position provider could also be useful. In that case we would need some kind of mechanism to specify a priority.

That is why I had split components before, it was implicitly picking the order in which the components were defined. This is one drawback to providing a single, simpler component.

Would you rather have a string array option to define the order? Like `[google, navigator, static]`?You still have the option to reorder/add a new provider with a hook on `onPositionResolved`.

> It might also be good to have a mechanism to have some kind of fallback position provider while we wait for a more accurate one

I could add (yet another) option like `useTemporaryPosition` to use the static values while waiting for another provider, but don't you think this would add another layer of complexity? :)

Reply:
Ok, we could have an option to `blockQueriesUntilPositionResolved` (Feel free to change the name :P ) , set to `true` by default.

Whenever it resolves, we allow the query to continue.

Even with the navigator provider (which is "slow" to resolve), just showing the loading animation while waiting for the user seems legit and logical.

Once again, is it really a real use case to configure multiple way to provide a location ? Or a way to hook "custom" location provider ? Why would someone really need that ?

Reply:
But say you want to call your own geolocation, 🤷‍♂️  fine. 

Can't use navigator if not in HTTPs, and you might not like to put your Google API key in the markup, so why not provide a way to call your own service.

We would use one for Sitecore, it has its own user-tracking which also has IP-geolocation service, but it is server-side, so we can use the `Static` provider with values from the model and it works 👌.

Reply:
Agreed, having as an `useTemporaryPosition` option, being able to add a custom geolocation provider and configuring the order of providers is probably overkill for now.

Reply:
The issue with the `getProvidersFromOptions()` method is it is dictating the order of providers as 1) Navigator, 2) Google, 3) Static. Also, we have 4 props (will be 5 with `useTemporaryPosition`?) in the options to handle the config, 'useNavigator', 'googleApiKey', 'longitudeValue', 'latitudeValue'. It is possible to simplify.

**A) If only one provider allowed**
We can use two properties: e.g. providerType: string and providerConfig: Object.
Depending on what providerType is set, the providerConfig object will need to hold the appropriate properties.
`getProvidersFromOptions()` will become a switch-case.

**B) If multiple providers allowed**
Same properties as (A) only now as arrays, which allows to set the order. I can see a benefit in multiple providers in that it allows attempting a TBD position via Navigator and/or Google before resorting to Static (or vice versa).

This will also make it easy to expand the provider list in the future.

Reply:
The scenario is `B) If multiple providers allowed`. 

What I don't like about the configuration list is the complexity that comes with it :)

Default usage is simple and will be good enough for most people. Advanced usage can still be achieved through event listeners.